lptree.pas

```diff
@@ -5652,11 +5668,34 @@ function TLapeTree_For.CompileBody(var Offset: Integer): TResVar;
   IncDec: TLapeTree_InternalMethod;
   tmpBody: TLapeTree_Base;
   tmpContinueStatements: TLapeFlowStatementList;
+  tmpVar: TResVar;
 begin
   Assert((FCondition <> nil) and (FCondition is TLapeTree_Operator));
   Assert((TLapeTree_Operator(FCondition).Right <> nil) and (TLapeTree_Operator(FCondition).Right is TLapeTree_ResVar));
 
   FStartBodyOffset := FCompiler.Emitter.CheckOffset(Offset);
+  
+  if LoopType = loopOver then
+  begin
+    FCounter.CompileToTempVar(Offset, tmpVar);
+
+    with TLapeTree_Operator.Create(op_Assign, FCompiler, @_DocPos) do
+    try
+      Left := TLapeTree_ResVar.Create(tmpVar.IncLock(), FCounter);
+      Right := TLapeTree_Operator.Create(op_Index, FCompiler);
+      with TLapeTree_Operator(Right) do
+      begin
+        Left := TLapeTree_ResVar.Create(Container.IncLock(), FLimit);
+        Right := TLapeTree_ResVar.Create(
+          TLapeTree_Operator(FCondition).Left.Compile(Offset).IncLock(), FCondition
+        );
+      end;
+      Compile(Offset);
+    finally
+      Free();
+    end;
+  end;
```

Comment:
Let's assume you where correct the other time (my first pr) you looked at this where you said I could not spill `tmpVar` here.